File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/hap/sign/SignBin.java
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.hap.sign;

import com.ohos.hapsigntool.hap.config.SignerConfig;
import com.ohos.hapsigntool.hap.entity.BlockHead;
import com.ohos.hapsigntool.hap.entity.SignHead;
import com.ohos.hapsigntool.hap.entity.SignContentInfo;
import com.ohos.hapsigntool.hap.entity.SignatureBlockTags;
import com.ohos.hapsigntool.hap.entity.SignatureBlockTypes;
import com.ohos.hapsigntool.error.SignatureException;
import com.ohos.hapsigntool.utils.FileUtils;
import com.ohos.hapsigntool.utils.HashUtils;
import com.ohos.hapsigntool.entity.ParamConstants;
import com.ohos.hapsigntool.utils.LogUtils;
import com.ohos.hapsigntool.utils.ParamProcessUtil;

import java.io.DataOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Map;

/**
 * LiteOS bin file Signature signer.
 *
 * @since 2021/12/21
 */
public class SignBin {
    private static final LogUtils LOGGER = new LogUtils(SignBin.class);

    /**
     * Constructor of Method
     */
    private SignBin() {
    }

    /**
     * Sign the bin file.
     *
     * @param signerConfig Config of the bin file to be signed.
     * @param signParams The input parameters of sign bin.
     * @return true if sign successfully; false otherwise.
     */
    public static boolean sign(SignerConfig signerConfig, Map<String, String> signParams) {
        /* 1. Make block head, write to output file. */
        String codesign = signParams.get(ParamConstants.PARAM_SIGN_CODE);
        if (ParamConstants.ProfileSignFlag.ENABLE_SIGN_CODE.getSignFlag().equals(codesign)) {
            LOGGER.warn("can not sign bin with codesign");
        }
        String inputFile = signParams.get(ParamConstants.PARAM_BASIC_INPUT_FILE);
        String outputFile = signParams.get(ParamConstants.PARAM_BASIC_OUTPUT_FILE);
        String profileFile = signParams.get(ParamConstants.PARAM_BASIC_PROFILE);
        String profileSigned = signParams.get(ParamConstants.PARAM_BASIC_PROFILE_SIGNED);
        if (!writeBlockDataToFile(inputFile, outputFile, profileFile, profileSigned)) {
            LOGGER.error("The block head data made failed.");
            ParamProcessUtil.delDir(new File(outputFile));
            return false;
        }
        LOGGER.info("The block head data made success.");

        /* 2. Make sign data, and write to output file */
        String signAlg = signParams.get(ParamConstants.PARAM_BASIC_SIGANTURE_ALG);
        if (!writeSignDataToOutputFile(signerConfig, outputFile, signAlg)) {
            LOGGER.error("The sign data made failed.");
            ParamProcessUtil.delDir(new File(outputFile));
            return false;
        }
        LOGGER.info("The data signed success.");
        /* 3. Make sign data, and write to output file */
        if (writeSignHeadDataToOutputFile(inputFile, outputFile)) {
            return true;
        }
        LOGGER.error("The sign head data made failed.");
        ParamProcessUtil.delDir(new File(outputFile));
        return false;
    }

    private static boolean writeBlockDataToFile(
            String inputFile, String outputFile, String profileFile, String profileSigned) {
        try {
            long binFileLen = FileUtils.getFileLen(inputFile);
            long profileDataLen = FileUtils.getFileLen(profileFile);
            if (!checkBinAndProfileLengthIsValid(binFileLen, profileDataLen)) {
                LOGGER.error("file length is invalid, binFileLen: " + binFileLen
                    + " profileDataLen: " + profileDataLen);
                throw new IOException();
            }

            long offset = binFileLen + BlockHead.getBlockLen() + BlockHead.getBlockLen();
            if (isLongOverflowInteger(offset)) {
                LOGGER.error("The profile block head offset is overflow integer range, offset: " + offset);
                throw new IOException();
            }
            char isSigned = SignatureBlockTypes.getProfileBlockTypes(profileSigned);
            byte[] proBlockByte =
                BlockHead.getBlockHead(isSigned, SignatureBlockTags.DEFAULT, (short) profileDataLen, (int) offset);

            offset += profileDataLen;
            if (isLongOverflowInteger(offset)) {
                LOGGER.error("The sign block head offset is overflow integer range, offset: " + offset);
                throw new IOException();
            }
            byte[] signBlockByte = BlockHead.getBlockHead(
                    SignatureBlockTypes.SIGNATURE_BLOCK, SignatureBlockTags.DEFAULT, (short) 0, (int) offset);

            return writeSignedBin(inputFile, proBlockByte, signBlockByte, profileFile, outputFile);
        } catch (IOException e) {
            LOGGER.error("writeBlockDataToFile failed.", e);
            return false;
        }
    }

    private static boolean writeSignedBin(String inputFile, byte[] proBlockByte, byte[] signBlockByte,
        String profileFile, String outputFile) {
        try (FileOutputStream fileOutputStream = new FileOutputStream(outputFile);
             DataOutputStream dataOutputStream = new DataOutputStream(fileOutputStream);) {
            // 1. write the input file to the output file.
            if (!FileUtils.writeFileToDos(inputFile, dataOutputStream)) {
                LOGGER.error("Failed to write information of input file: " + inputFile +
                        " to outputFile: " + outputFile);
                throw new IOException();
            }

            // 2. write profile block head to the output file.
            if (!FileUtils.writeByteToDos(proBlockByte, dataOutputStream)) {
                LOGGER.error("Failed to write proBlockByte to output file: " + outputFile);
                throw new IOException();
            }

            // 3. write sign block head to the output file.
            if (!FileUtils.writeByteToDos(signBlockByte, dataOutputStream)) {
                LOGGER.error("Failed to write binBlockByte to output file: " + outputFile);
                throw new IOException();
            }

            // 4. write profile src file to the output file.
            if (!FileUtils.writeFileToDos(profileFile, dataOutputStream)) {
                LOGGER.error("Failed to write profile file: " + profileFile);
                throw new IOException();
            }
        } catch (IOException e) {
            LOGGER.error("writeSignedBin failed.", e);
            return false;
        }
        return true;
    }

    private static boolean checkBinAndProfileLengthIsValid(long binFileLen, long profileDataLen) {
        return (binFileLen != -1) && (profileDataLen != -1) && (!isLongOverflowShort(profileDataLen));
    }

    private static boolean writeSignHeadDataToOutputFile(String inputFile, String outputFile) {
        long size = FileUtils.getFileLen(outputFile) - FileUtils.getFileLen(inputFile) + SignHead.SIGN_HEAD_LEN;
        if (isLongOverflowInteger(size)) {
            LOGGER.error("File size is overflow integer range.");
            return false;
        }
        SignHead signHeadData = new SignHead();
        byte[] signHeadByte = signHeadData.getSignHead((int) size);
        if (signHeadByte == null || signHeadByte.length == 0) {
            LOGGER.error("Failed to get sign head data.");
            return false;
        }
        return FileUtils.writeByteToOutFile(signHeadByte, outputFile);
    }

    private static boolean writeSignDataToOutputFile(SignerConfig signerConfig, String outputFile, String signAlg) {
        String alg = ParamProcessUtil.getSignatureAlgorithm(signAlg).getContentDigestAlgorithm().getDigestAlgorithm();
        byte[] data = HashUtils.getFileDigest(outputFile, alg);
        if (data == null) {
            LOGGER.error("getFileDigest failed.");
            return false;
        }
        byte[] outputChunk = null;
        SignContentInfo contentInfo = null;
        try {
            contentInfo = new SignContentInfo();
            contentInfo.addContentHashData(
                    (char) 0, SignatureBlockTags.HASH_ROOT_4K, (short) HashUtils.getHashAlgsId(alg), data.length, data);
            byte[] dig = contentInfo.getByteContent();
            outputChunk = Pkcs7Generator.BC.generateSignedData(dig, signerConfig);
            return FileUtils.writeByteToOutFile(outputChunk, outputFile);
        } catch (SignatureException e) {
            LOGGER.error("Sign hap Lite failed.", e);
        }
        return false;
    }

    private static boolean isLongOverflowInteger(long num) {
        return (num - (num & 0xffffffffL)) != 0;
    }

    private static boolean isLongOverflowShort(long num) {
        return (num - (num & 0xffffL)) != 0;
    }
}

File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/hap/sign/src/sign_bin.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "sign_bin.h"
#include "param_constants.h"
#include "file_utils.h"
#include "block_head.h"
#include "signature_block_types.h"
#include "signature_block_tags.h"
#include "hash_utils.h"
#include "sign_content_info.h"
#include "sign_head.h"
#include "bc_pkcs7_generator.h"
#include "params.h"

namespace OHOS {
namespace SignatureTools {

bool SignBin::Sign(SignerConfig& signerConfig, const std::map<std::string, std::string>& signParams)
{
    /* 1. Make block head, write to output file. */
    std::string signCode = signParams.at(ParamConstants::PARAM_SIGN_CODE);
    if (ParamConstants::ENABLE_SIGN_CODE == signCode) {
        SIGNATURE_TOOLS_LOGW("can not sign bin with codesign.\n");
    }
    std::string inputFile = signParams.at(ParamConstants::PARAM_BASIC_INPUT_FILE);
    std::string outputFile = signParams.at(ParamConstants::PARAM_BASIC_OUTPUT_FILE);
    std::string profileFile = signParams.at(ParamConstants::PARAM_BASIC_PROFILE);
    std::string profileSigned = signParams.at(ParamConstants::PARAM_BASIC_PROFILE_SIGNED);
    bool writeBlockOk = WriteBlockDataToFile(inputFile, outputFile, profileFile, profileSigned);
    if (!writeBlockOk) {
        SIGNATURE_TOOLS_LOGE("The block head data made failed.");
        FileUtils::DelDir(outputFile);
        return false;
    }
    /* 2. Make sign data, and append write to output file */
    std::string signAlg = signParams.at(ParamConstants::PARAM_BASIC_SIGANTURE_ALG);
    bool writeSignDataOk = WriteSignDataToOutputFile(signerConfig, outputFile, signAlg);
    if (!writeSignDataOk) {
        SIGNATURE_TOOLS_LOGE("The sign data made failed.");
        FileUtils::DelDir(outputFile);
        return false;
    }
    /* 3. Make sign head data, and write to output file */
    bool writeSignHeadDataOk = WriteSignHeadDataToOutputFile(inputFile, outputFile);
    if (!writeSignHeadDataOk) {
        SIGNATURE_TOOLS_LOGE("The sign head data made failed.");
        FileUtils::DelDir(outputFile);
        return false;
    }
    return true;
}

bool SignBin::WriteBlockDataToFile(const std::string& inputFile, const std::string& outputFile,
                                   const std::string& profileFile, const std::string& profileSigned)
{
    int64_t binFileLen = FileUtils::GetFileLen(inputFile);
    int64_t profileDataLen = FileUtils::GetFileLen(profileFile);
    bool isValid = CheckBinAndProfileLengthIsValid(binFileLen, profileDataLen);
    if (!isValid) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR,
                            "file length is invalid, binFileLen: " + std::to_string(binFileLen) +
                            "lld, profileDataLen: " + std::to_string(profileDataLen) + "lld");
        return false;
    }
    int64_t offset = binFileLen + BlockHead::GetBlockLen() + BlockHead::GetBlockLen();
    bool isOver = IsLongOverflowInteger(offset);
    if (isOver) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR,
            "The profile block head offset is overflow interger range in sign bin, offset: "
            + std::to_string(offset));
        return false;
    }
    char isSigned = SignatureBlockTypes::GetProfileBlockTypes(profileSigned);
    std::string proBlockByte =
        BlockHead::GetBlockHead(isSigned, SignatureBlockTags::DEFAULT, (short)profileDataLen, (int)offset);
    offset += profileDataLen;
    isOver = IsLongOverflowInteger(offset);
    if (isOver) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR,
            "The profile block head offset is overflow interger range in sign bin, offset: "
            + std::to_string(offset));
        return false;
    }
    std::string signBlockByte = BlockHead::GetBlockHead(
        SignatureBlockTypes::SIGNATURE_BLOCK, SignatureBlockTags::DEFAULT, (short)0, (int)offset);
    return WriteSignedBin(inputFile, proBlockByte, signBlockByte, profileFile, outputFile);
}

std::vector<int8_t> SignBin::GenerateFileDigest(const std::string& outputFile,
                                                const std::string& signAlg)
{
    SignatureAlgorithmHelper signatureAlgorithmClass;
    bool getAlgOk = Params::GetSignatureAlgorithm(signAlg, signatureAlgorithmClass);
    if (!getAlgOk) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignHap] get Signature Algorithm failed.");
        return std::vector<int8_t>();
    }
    std::string alg = signatureAlgorithmClass.m_contentDigestAlgorithm.GetDigestAlgorithm();
    std::vector<int8_t> data = HashUtils::GetFileDigest(outputFile, alg);
    if (data.empty()) {
        SIGNATURE_TOOLS_LOGE("GetFileDigest failed.");
        return std::vector<int8_t>();
    }
    SignContentInfo contentInfo;
    contentInfo.AddContentHashData(0, SignatureBlockTags::HASH_ROOT_4K, HashUtils::GetHashAlgsId(alg),
                                   data.size(), data);
    return contentInfo.GetByteContent();
}

bool SignBin::WriteSignDataToOutputFile(SignerConfig& SignerConfig, const std::string& outputFile,
                                        const std::string& signAlg)
{
    std::vector<int8_t> dig = GenerateFileDigest(outputFile, signAlg);
    if (dig.empty()) {
        SIGNATURE_TOOLS_LOGE("generateSignature verity digest is null.");
        return false;
    }
    std::string retStr;
    std::string signedData(dig.begin(), dig.end());
    std::unique_ptr<Pkcs7Generator> generator = std::make_unique<BCPkcs7Generator>();
    if (generator->GenerateSignedData(signedData, &SignerConfig, retStr)) {
        SIGNATURE_TOOLS_LOGE("failed to GenerateSignedData!");
        return false;
    }
    bool writeByteToOutFile = FileUtils::AppendWriteByteToFile(retStr, outputFile);
    if (!writeByteToOutFile) {
        SIGNATURE_TOOLS_LOGE("write signedData to outputFile failed!");
        return false;
    }
    return true;
}

bool SignBin::WriteSignHeadDataToOutputFile(const std::string& inputFile, const std::string& outputFile)
{
    int64_t size = FileUtils::GetFileLen(outputFile) - FileUtils::GetFileLen(inputFile) + SignHead::SIGN_HEAD_LEN;
    bool isOver = IsLongOverflowInteger(size);
    if (isOver) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR,
            "File size is Overflow integer range, size: " + std::to_string(size));
        return false;
    }
    SignHead signHeadData;
    std::vector<int8_t> signHeadByte = signHeadData.GetSignHead(size);
    if (signHeadByte.empty()) {
        SIGNATURE_TOOLS_LOGE("Failed to get sign head data!");
        return false;
    }
    bool writeByteToOutFile =
        FileUtils::AppendWriteByteToFile(std::string(signHeadByte.begin(), signHeadByte.end()), outputFile);
    if (!writeByteToOutFile) {
        SIGNATURE_TOOLS_LOGE("Failed to WriteByteToOutFile!");
        return false;
    }
    return true;
}

bool SignBin::CheckBinAndProfileLengthIsValid(int64_t binFileLen, int64_t profileDataLen)
{
    return binFileLen != -1 && profileDataLen != -1 && !IsLongOverflowShort(profileDataLen);
}

bool SignBin::IsLongOverflowInteger(int64_t num)
{
    return (num - (num & 0xffffffffL)) != 0;
}

bool SignBin::IsLongOverflowShort(int64_t num)
{
    return (num - (num & 0xffff)) != 0;
}

bool SignBin::WriteSignedBin(const std::string& inputFile, const std::string& proBlockByte,
                             const std::string& signBlockByte, const std::string& profileFile,
                             const std::string& outputFile)
{
    // 1. write the input file to the output file.
    bool writeInputOk = FileUtils::WriteInputToOutPut(inputFile, outputFile);
    // 2. append write profile block head to the output file.
    bool appendProfileHeadOk = FileUtils::AppendWriteByteToFile(proBlockByte, outputFile);
    // 3. append write sign block head to the output file.
    bool appendBlockHeadOk = FileUtils::AppendWriteByteToFile(signBlockByte, outputFile);
    // 4. write profile src file to the output file.
    bool appendProfileSrcOk = FileUtils::AppendWriteFileToFile(profileFile, outputFile);
    if (!writeInputOk || !appendProfileHeadOk || !appendBlockHeadOk || !appendProfileSrcOk) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "Failed to write signed bin");
        return false;
    }
    return true;
}

} // namespace SignatureTools
} // namespace OHOS